### PR TITLE
Allow underscores in numeric literals

### PR DESCRIPTION
**Pull request by seriyps**
Just draft to get some feedback. Can update docs, erlang-mode.el and tests if we are ok to proceed.

This is to make long numeric literals more readable. The idea is [borrowed from Rust language](https://doc.rust-lang.org/rust-by-example/primitives/literals.html) ([rust spec](https://doc.rust-lang.org/reference/tokens.html#numbers))

Examples:
* `123_456_789` vs `123456789`
* `123_456.789_123` vs `123456.789123`
* `16#123_ABC` vs `16#123ABC`
* `2#1100_0011_0101` vs `2#110000110101`

<del>One minor problem with current implementation is that `erl_scan:string("123_456", text)` will return `{integer,[{text,"123456"},<..>],123456}` - so, text will be "123456" and not original "123_456", but this can be fixed, if it's important.</del>

### CODE REVIEW

**Comment by mikpe** [comment]
I welcome this in principle.  The digit separator feature has been around since at least Ada83, and is currently present in about 7 or 8 additional languages according to a proposal to add them to "Go".

I would suggest to not allow $_ wherever a digit is expected, but only as a separator between digits (the Ada rule).

See <https://go.googlesource.com/proposal/+/master/design/19308-number-literals.md> section "Digit Separators" for some background.

**Comment by seriyps** [comment]
@mikpe I also think it doesn't makes much sense to support this in base. In current implementation you can have `1_6#12F` just because it was easier to implement.

> I would suggest to not allow $_ wherever a digit is expected, but only as a separator between digits (the Ada rule).

So, only allow it if chars left and right from `$_` are digits?

With current implementation all of this should work, but doesn't make much sense:

* `123___456` (more than one underscore)
* `123_.456`
* `123._456`
* `1_6#12F`
* `16#_12F`
* `123_`
* `123.456_`
* `16#12F_`

Leading underscore `_12345` will be interpreted as variable name.

**Comment by mikpe** [comment]
Yes, leading (_123), trailing (123_), or repeated (12__34) underscores should not be allowed.  That's what's meant by being a separator between groups of digits.

In Erlang the base is simply a decimal numeral, and I would prefer to allow underscores there for consistency -- I don't think it causes any problems.

**Comment by seriyps** [comment]
@mikpe I updated my implementation according to your comments (as a separate commit for now).

File that I was using to test it:

```
-module(mydigits).

-export([ok/0, fail/0]).

ok() ->
    [123_456,
     123_456_789,
     123_456.789,
     123.456_789,
     1.2_345e10,
     1.234e1_06,
     16#123_ABC_456_DEF_789,
     2#0011_0101_0011,
     1_6#123ABC].

fail() ->
    Results = lists:map(
      fun erl_scan:string/1,
      [
       "123_",
       "123_456_",
       "123__456",
       "123_.456",
       "123._456",
       "123.456_",
       "123._",
       "1._23e10",
       "1.23e_10",
       "1.23e10_",
       "16_#123ABC",
       "16#123_",
       "16#_123",
       "16#ABC_",
       "16#_ABC",
       "2#_0101",
       "2#0101_"
      ]),
    {Results,
     lists:filter(fun({ok, [{Tok, _, _}]}) ->
                       (Tok == integer) orelse (Tok == float);
		     (_) ->
		       false
		  end, Results)}.
```

File: lib/stdlib/src/erl_scan.erl [hunk 1]
```diff
@@ -940,13 +940,16 @@ escape_char(C) -> C.
 
 scan_number([C|Cs], St, Line, Col, Toks, Ncs) when ?DIGIT(C) ->
     scan_number(Cs, St, Line, Col, Toks, [C|Ncs]);
+scan_number([$_,Next|Cs], St, Line, Col, Toks, [Prev|_]=Ncs) when
+      ?DIGIT(Next) andalso ?DIGIT(Prev) ->
```
Comment (seriyps) [inline lib/stdlib/src/erl_scan.erl]: might be more efficient to rewrite (here and in other cases) as

```
scan_number([Pre, $_, Post | Cs], ...) when ?DIGIT(Pre), ?DIGIT(Post) ->
    scan_number(..., [Post, $_, Pre | Ncs]);
```

and put as 1st function clause.

Should I?

Comment (mikpe) [reply lib/stdlib/src/erl_scan.erl]: I think that would pessimize the common case (no $_), while not really optimizing the $_ case.  The current code looks fine to me.

**Comment by seriyps** [comment]
I guess it makes sense to update documentation as well. At least here: http://erlang.org/doc/reference_manual/data_types.html#number
Any other suggestions?

**Comment by uabboli** [comment]
We've had a Technical Board (TB) meeting discussing your PR, and the
bottom line is that we think it's a nice thing to have, and that it's
worth the effort to finish it up. We hope you are willing to put in
some more work. What follows is a summary of the conclusions of the
meeting.

Since it's a language extension, we think there should be an EEP (see
https://github.com/erlang/eep).

We agree with @mikpe: underscore is to be allowed between digits (as a
consequence repeated underscores are not allowed). They are to be
allowed in the base of integer numbers and the exponent of floats, if
that makes the implementation simpler.

The underscores are to be lost when representing abstract forms or
terms as strings (erl_pp, io_lib).

As you suggest, the Erlang Reference Manuals needs to be updated
(section 3.2 Number).

The Erlang mode for Emacs needs to be updated.

The Erlang Debugger uses wxWidgets, which in turn uses Scintilla. It
is not a requirement for this PR to update the lexer used in Scintilla.

Beside these general comments from the TB, I've looked at your code
and fixed a bug. When at it, I've also optimized the code a bit. I'll
see if I can update your branch with my changes (3 commits).

**Comment by seriyps** [comment]
> we think it's a nice thing to have

Great news, thanks!

> Since it's a language extension, we think there should be an EEP (see https://github.com/erlang/eep).

Oh, I will try, yes.

> Erlang Reference Manuals needs to be updated (section 3.2 Number)

Will do

> The Erlang mode for Emacs needs to be updated.

The only place I found in erlang.el is:

https://github.com/erlang/otp/blob/a62aed81c56c724f7dd7040adecaa28a78e5d37f/lib/tools/emacs/erlang.el#L1227-L1230

I guess it's responsible for highlighting `16#af_56` as numeric constant and not record? Will look at it, but please let me know if I'm missing something else.

Should I care about `fun my_mod:my_fun/12_34`? (Apparently, this syntax is also now supported with this PR).

> and fixed a bug

a-ha! That's why I was not able to use `test_string` in tests. Now I see. Thanks!

**Comment by uabboli** [comment]
> I guess it's responsible for highlighting 16#af_56 as numeric constant
> and not record? Will look at it, but please let me know if I'm missing
> something else.
>
> Should I care about fun my_mod:my_fun/12_34? (Apparently, this syntax
> is also now supported with this PR).

I have to admit that I don't know much about the Emacs mode. If you
can fix the simple cases I think it's good enough for this PR. We at
OTP don't have much time to help with the Emacs mode, so if you run
into trouble, we'll have to rely on contributions from more informed
people (as we've done for quite some time now).

**Comment by seriyps** [comment]
EEP was submitter to mailing list as well as in a form of PR to eep github repository.

**Comment by seriyps** [comment]
I updated the docs and emacs erlang-mode. Is there anything I can do now?

**Comment by uabboli** [comment]
I think it looks good.

I cannot make the emacs mode work. In fact we (I and Dan) cannot see
any improvement. I think it's better to leave erlang.el without any changes,
unless you can show that the modifications actually make a difference.

The examples in the docs are not colored correctly. I've been
informed that the program we use can be found at
https://highlightjs.org/.

My take on this PR is that we can manage without updated highlightings
(Emacs, Debugger, docs). Later contributions are welcome, of course.

Thank you for your effort!

**Comment by seriyps** [comment]
> I cannot make the emacs mode work. In fact we (I and Dan) cannot see
any improvement. I think it's better to leave erlang.el without any changes,
unless you can show that the modifications actually make a difference.

The only difference for emacs mode is how it highlights `16#abcd_1234`: without this modification it will highlight `_1234` as a record and with modification it won't apply highlighting.
before:
image

after:

image

So, if you think emacs part should be skipped, should I remove emacs-related commit or you will take care of it?

**Comment by seriyps** [comment]
regarding highlightjs, yep, it would be relatively easy to fix in the upstream:

https://github.com/highlightjs/highlight.js/blob/2073d6b314aa63d44007e883194c27bcd0859015/src/languages/erlang.js#L23

https://github.com/highlightjs/highlight.js/blob/2073d6b314aa63d44007e883194c27bcd0859015/src/languages/erlang-repl.js#L25

but then the version of the lib that is bundled needs to be updated in Erlang as well like it was done here https://github.com/erlang/otp/pull/2421/commits/5e30a839258f2685fb064f3e1ec381766df94175

but it will be a separate PR i guess

**Comment by uabboli** [comment]
> So, if you think emacs part should be skipped, should I remove
> emacs-related commit or you will take care of it?

I'll take of it.

Again, thanks for your effort!

**Pull request merged**